### ISSUE TITLE
landscape/reverse-landscape broken on macOS

### ISSUE THREAD

**Issue by klabarge** [qzind/tray#197]
**OS:** macOS 10.12.4
**QZ Tray:** 2.0.4-RC.1
**Browser:** Tested on Chrome and Safari

On macOS, `orientation: landscape` and `orientation: reverse-landscape`, images are rotated the same amount. 

**UPDATE:** I believe the HTML and Image printing issue is separate.
> This effects HTML, PDF, and Image printing.

**landscape stack trace**
```
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"PDFwriter"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":null,"interpolation":"","jobName":"","margins":"","orientation":"landscape","paperThickness":"","printerTray":"","rasterize":true,"rotation":"","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"image","data":"file:///Applications/QZ%20Tray.app/demo/assets/img/image_sample.png"}]},"timestamp":1494990391982,"uid":"rjcsqv"}
[WARN] 2017-05-16 23:06:31,983 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2017-05-16 23:06:33,099 @ qz.common.TrayManager:?
	Allowed localhost to print to PDFwriter
[TRACE] 2017-05-16 23:06:33,099 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/3 already in use
[DEBUG] 2017-05-16 23:06:33,100 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintImage to print
[DEBUG] 2017-05-16 23:06:33,100 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching PDFwriter
[DEBUG] 2017-05-16 23:06:33,100 @ qz.printer.PrintServiceMatcher:?
	Found 3 printers
[DEBUG] 2017-05-16 23:06:33,110 @ qz.printer.PrintServiceMatcher:?
	Found match: PDFwriter
[WARN] 2017-05-16 23:06:33,110 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2017-05-16 23:06:33,110 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2017-05-16 23:06:33,111 @ qz.printer.PrintOptions:?
	Cannot read  as a double for margins, using default
[WARN] 2017-05-16 23:06:33,111 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[WARN] 2017-05-16 23:06:33,111 @ qz.printer.PrintOptions:?
	Cannot read  as a double for rotation, using default
[DEBUG] 2017-05-16 23:06:33,116 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra, -l]
[DEBUG] 2017-05-16 23:06:33,225 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra_class, -l]
[DEBUG] 2017-05-16 23:06:33,251 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PDFwriter, -l]
[DEBUG] 2017-05-16 23:06:33,274 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Zebra_epl, -l]
[WARN] 2017-05-16 23:06:33,294 @ qz.printer.PrintOptions:?
	Cannot read null as a double for fallbackDensity, using default
[DEBUG] 2017-05-16 23:06:33,315 @ qz.printer.action.PrintImage:?
	Parsed 1 images for printing
[TRACE] 2017-05-16 23:06:33,325 @ qz.printer.action.PrintPixel:?
	DPI: 600.0	CNV: 72.0
[TRACE] 2017-05-16 23:06:33,325 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2017-05-16 23:06:33,325 @ qz.printer.action.PrintPixel:?
	[60000x60000 dphi, color, landscape, (0.0,0.0)->(215.9,279.4)mm]
[INFO] 2017-05-16 23:06:33,333 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-05-16 23:06:33,333 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-05-16 23:06:33,367 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-05-16 23:06:33,368 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-05-16 23:06:33,369 @ qz.printer.action.PrintImage:?
	Scaling image up by x8.333333333333334
[DEBUG] 2017-05-16 23:06:33,482 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:792,612
[TRACE] 2017-05-16 23:06:33,482 @ qz.printer.action.PrintImage:?
	Image size: 792.0,316.6098439375751
[TRACE] 2017-05-16 23:06:33,482 @ qz.printer.action.PrintImage:?
	{Alpha blending interpolation method key=Highest quality alpha blending methods, Global antialiasing enable key=Antialiased rendering mode, Stroke normalization control key=Pure stroke conversion for accurate paths, Image interpolation method key=Bicubic image interpolation mode, Global rendering quality key=Highest quality rendering methods, Color rendering quality key=Highest quality color rendering mode, Text-specific LCD contrast key=140, Fractional metrics enable key=Integer text metrics mode, Text-specific antialiasing enable key=Default antialiasing text mode}
[DEBUG] 2017-05-16 23:06:33,482 @ qz.printer.action.PrintImage:?
	Memory: 907m/1820m
[INFO] 2017-05-16 23:06:33,555 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2017-05-16 23:06:33,556 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**reverse-landscape stack trace**
```
Message: {"call":"print","promise":{},"params":{"printer":{"name":"PDFwriter"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":null,"interpolation":"","jobName":"","margins":"","orientation":"reverse-landscape","paperThickness":"","printerTray":"","rasterize":true,"rotation":"","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"image","data":"file:///Applications/QZ%20Tray.app/demo/assets/img/image_sample.png"}]},"timestamp":1494990365121,"uid":"idkoac"}
[WARN] 2017-05-16 23:06:05,148 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2017-05-16 23:06:06,744 @ qz.common.TrayManager:?
	Allowed localhost to print to PDFwriter
[TRACE] 2017-05-16 23:06:06,745 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/3 already in use
[DEBUG] 2017-05-16 23:06:06,745 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintImage to print
[DEBUG] 2017-05-16 23:06:06,746 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching PDFwriter
[DEBUG] 2017-05-16 23:06:06,746 @ qz.printer.PrintServiceMatcher:?
	Found 3 printers
[DEBUG] 2017-05-16 23:06:06,764 @ qz.printer.PrintServiceMatcher:?
	Found match: PDFwriter
[WARN] 2017-05-16 23:06:06,765 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2017-05-16 23:06:06,765 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2017-05-16 23:06:06,765 @ qz.printer.PrintOptions:?
	Cannot read  as a double for margins, using default
[WARN] 2017-05-16 23:06:06,765 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[WARN] 2017-05-16 23:06:06,765 @ qz.printer.PrintOptions:?
	Cannot read  as a double for rotation, using default
[DEBUG] 2017-05-16 23:06:06,769 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra, -l]
[DEBUG] 2017-05-16 23:06:06,892 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, zebra_class, -l]
[DEBUG] 2017-05-16 23:06:06,912 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PDFwriter, -l]
[DEBUG] 2017-05-16 23:06:06,933 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Zebra_epl, -l]
[WARN] 2017-05-16 23:06:06,954 @ qz.printer.PrintOptions:?
	Cannot read null as a double for fallbackDensity, using default
[DEBUG] 2017-05-16 23:06:07,004 @ qz.printer.action.PrintImage:?
	Parsed 1 images for printing
[TRACE] 2017-05-16 23:06:07,011 @ qz.printer.action.PrintPixel:?
	DPI: 600.0	CNV: 72.0
[TRACE] 2017-05-16 23:06:07,011 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2017-05-16 23:06:07,011 @ qz.printer.action.PrintPixel:?
	[60000x60000 dphi, color, reverse-landscape, (0.0,0.0)->(215.9,279.4)mm]
[INFO] 2017-05-16 23:06:07,019 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-05-16 23:06:07,019 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-05-16 23:06:07,141 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-05-16 23:06:07,142 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-05-16 23:06:07,144 @ qz.printer.action.PrintImage:?
	Scaling image up by x8.333333333333334
[DEBUG] 2017-05-16 23:06:07,252 @ qz.printer.action.PrintImage:?
	Paper area: 0,0:792,612
[TRACE] 2017-05-16 23:06:07,252 @ qz.printer.action.PrintImage:?
	Image size: 792.0,316.6098439375751
[TRACE] 2017-05-16 23:06:07,253 @ qz.printer.action.PrintImage:?
	{Alpha blending interpolation method key=Highest quality alpha blending methods, Global antialiasing enable key=Antialiased rendering mode, Stroke normalization control key=Pure stroke conversion for accurate paths, Image interpolation method key=Bicubic image interpolation mode, Global rendering quality key=Highest quality rendering methods, Color rendering quality key=Highest quality color rendering mode, Text-specific LCD contrast key=140, Fractional metrics enable key=Integer text metrics mode, Text-specific antialiasing enable key=Default antialiasing text mode}
[DEBUG] 2017-05-16 23:06:07,253 @ qz.printer.action.PrintImage:?
	Memory: 895m/1820m
[INFO] 2017-05-16 23:06:07,550 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2017-05-16 23:06:07,550 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**Comment by tresf** [comment]
I have a suspicion this is related to https://github.com/qzind/tray/issues/196.

**Comment by klabarge** [comment]
There seems to be several bugs related to `orientation`. I will break this down between each OS. Let me know if you would like this broken out into separate bug reports.

### macOS
1. Image and HTML `reverse-landscape` is broken

   I went as far back as 2.0.1 and I could not get `reverse-landscape` to work on macOS for Image and HTML printing. The `orientation` remains `landscape`.

1. PDF `reverse-landscape` is broken

   With the latest PDF changes ([snapshotted here with 2.0.4-RC.1]( https://github.com/klabarge/tray/releases/tag/v2.0.4-RC.1)), `reverse-landscape`does not work, `orientation` remains `landscape`.

### Linux

1. `landscape` does not work 

   With the latest PDF changes ([snapshotted here with 2.0.4-RC.1]( https://github.com/klabarge/tray/releases/tag/v2.0.4-RC.1)), `landscape` does not work. The `orientation` remains `portrait` with both `landscape` and `reverse-landscape` printing.

### Windows

I have not surfaced any Windows-specific issues.

### All OS's
1. inconsistent  PDF `landscape` printing. 

   PDF printing, as @bberenz noticed/documented here: https://github.com/qzind/tray/issues/196#issuecomment-310771992 has worked differently throughout different versions.

**Comment by bberenz** [comment]
> **Linux**
> 1. `landscape` does not work

Can you provide logs, testing off the now pdffix branch, with the latest pdf/orientation changes, I do not see this in my vm (Ubuntu 16.04)

**Comment by klabarge** [comment]
> Can you provide logs, testing off the now pdffix branch, with the latest pdf/orientation changes, I do not see this in my vm (Ubuntu 16.04)

Sorry @bberenz, I overlooked that my PDF Viewer changes the properties to `11.0 X 8.5in`. `landscape` does work.

![image](https://user-images.githubusercontent.com/12505463/27980926-fe8dd9ba-6351-11e7-9cb0-65de17fe6873.png)

Did you test `reverse-landscape`? This seems to be identical as `landscape`.

**Comment by bberenz** [comment]
The lack of difference in reverse landscape might be related to the viewer you are using, as I see the same thing with cups PDF. Have you tried printing to a physical printer?

Here's my print to a physical printer from an ubuntu16 VM:
![physical](https://user-images.githubusercontent.com/5912715/28027096-fe0922ea-6565-11e7-9088-f67981faefe0.jpg)
I drew the arrows on the paper before printing to be sure of the direction

**Comment by klabarge** [comment]
I use the default PDF viewer in Ubuntu, and I also downloaded [Foxit PDF-Reader](https://www.foxitsoftware.com/pdf-reader/) with the same results.

I will test to a physical printer and report back.

**Comment by klabarge** [comment]
I did testing to a physical printer (HP LaserJet P4515) using the pdf-fix branch on all OS's.

@bberenz, you are right - `landscape` and `reverse-landscape` printing work in Ubuntu. Both Windows and Ubuntu print identically.

I reproduced two issues on macOS:

1. `landscape` and `reverse-landscape` are the same on macOS.

1. `landscape` on macOS is printed the same as `reverse-landscape` on Windows and Linux. It is oriented opposite of these platforms.

| |landscape| reverse-landscape |
|---|---|---|
|mac | ⛔️  | ⛔️  |
|Windows| ✅ | ✅ 
|linux| ✅ | ✅ |


#### Reverse landscape - macOS
```
[DEBUG] 2017-07-24 17:26:39,586 @ qz.ws.PrintSocketClient:?
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"10.10.0.50"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":null,"interpolation":"","jobName":"REVERSE","margins":"0","orientation":"reverse-landscape","paperThickness":"","printerTray":"","rasterize":true,"rotation":"0","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"image","data":"file:///Applications/QZ%20Tray.app/demo/assets/img/image_sample.png"}]},"timestamp":1500931599585,"uid":"rlatv7"}
[WARN] 2017-07-24 17:26:39,590 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2017-07-24 17:26:41,007 @ qz.common.TrayManager:?
	Allowed localhost to print to 10.10.0.50
[TRACE] 2017-07-24 17:26:41,008 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/5 already in use
[DEBUG] 2017-07-24 17:26:41,009 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintImage to print
[DEBUG] 2017-07-24 17:26:41,009 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching 10.10.0.50
[DEBUG] 2017-07-24 17:26:41,011 @ qz.printer.PrintServiceMatcher:?
	Found 7 printers
[DEBUG] 2017-07-24 17:26:41,022 @ qz.printer.PrintServiceMatcher:?
	Found match: 10.10.0.50
[WARN] 2017-07-24 17:26:41,023 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2017-07-24 17:26:41,023 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2017-07-24 17:26:41,024 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[DEBUG] 2017-07-24 17:26:41,027 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_54, -l]
[DEBUG] 2017-07-24 17:26:41,045 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS _10_10_0_54: 300dpi
[DEBUG] 2017-07-24 17:26:41,046 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Canon_MX700_series, -l]
[DEBUG] 2017-07-24 17:26:41,083 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS Canon_MX700_series: 600dpi
[DEBUG] 2017-07-24 17:26:41,084 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_53, -l]
[DEBUG] 2017-07-24 17:26:41,120 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PRTLASER3, -l]
[DEBUG] 2017-07-24 17:26:41,152 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_50, -l]
[DEBUG] 2017-07-24 17:26:41,172 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Brother_PT_2730, -l]
[DEBUG] 2017-07-24 17:26:41,193 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS Brother_PT_2730: 180dpi
[DEBUG] 2017-07-24 17:26:41,194 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PRTLASER, -l]
[DEBUG] 2017-07-24 17:26:41,225 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS PRTLASER: 600dpi
[WARN] 2017-07-24 17:26:41,226 @ qz.printer.PrintOptions:?
	Cannot read null as a double for fallbackDensity, using default
[DEBUG] 2017-07-24 17:26:41,248 @ qz.printer.action.PrintImage:?
	Parsed 1 images for printing
[TRACE] 2017-07-24 17:26:41,263 @ qz.printer.action.PrintPixel:?
	DPI: 600.0	CNV: 72.0
[TRACE] 2017-07-24 17:26:41,264 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2017-07-24 17:26:41,264 @ qz.printer.action.PrintPixel:?
	[reverse-landscape, color, 60000x60000 dphi, (0.0,0.0)->(215.9,279.4)mm]
[INFO] 2017-07-24 17:26:41,274 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-07-24 17:26:41,274 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-07-24 17:26:41,336 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-07-24 17:26:41,337 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-07-24 17:26:41,338 @ qz.printer.action.PrintImage:?
	Scaling image up by x8.333333333333334
[DEBUG] 2017-07-24 17:26:41,456 @ qz.printer.action.PrintImage:?
	Paper area: 18,16:754,578
[TRACE] 2017-07-24 17:26:41,456 @ qz.printer.action.PrintImage:?
	Image size: 754.800048828125,301.7387950297307
[TRACE] 2017-07-24 17:26:41,457 @ qz.printer.action.PrintImage:?
	{Fractional metrics enable key=Integer text metrics mode, Image interpolation method key=Bicubic image interpolation mode, Global antialiasing enable key=Antialiased rendering mode, Color rendering quality key=Highest quality color rendering mode, Alpha blending interpolation method key=Highest quality alpha blending methods, Text-specific antialiasing enable key=Default antialiasing text mode, Stroke normalization control key=Pure stroke conversion for accurate paths, Global rendering quality key=Highest quality rendering methods, Text-specific LCD contrast key=140}
[DEBUG] 2017-07-24 17:26:41,460 @ qz.printer.action.PrintImage:?
	Memory: 894m/2731m
[INFO] 2017-07-24 17:26:41,522 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2017-07-24 17:26:41,523 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

#### Landscape - macOS
```
[DEBUG] 2017-07-24 17:26:49,947 @ qz.ws.PrintSocketClient:?
	Message: {"call":"print","promise":{},"params":{"printer":{"name":"10.10.0.50"},"options":{"colorType":"color","copies":"1","density":"","duplex":false,"fallbackDensity":null,"interpolation":"","jobName":"LANDSCAPE","margins":"0","orientation":"landscape","paperThickness":"","printerTray":"","rasterize":true,"rotation":"0","scaleContent":true,"size":null,"units":"in","altPrinting":false,"encoding":"","endOfDoc":"","perSpool":"1"},"data":[{"type":"image","data":"file:///Applications/QZ%20Tray.app/demo/assets/img/image_sample.png"}]},"timestamp":1500931609947,"uid":"df3i9r"}
[WARN] 2017-07-24 17:26:49,950 @ qz.ws.PrintSocketClient:?
	Bad signature on request
[INFO] 2017-07-24 17:26:51,751 @ qz.common.TrayManager:?
	Allowed localhost to print to 10.10.0.50
[TRACE] 2017-07-24 17:26:51,752 @ qz.utils.PrintingUtilities:?
	Waiting for processor, 0/5 already in use
[DEBUG] 2017-07-24 17:26:51,752 @ qz.utils.PrintingUtilities:?
	Using qz.printer.action.PrintImage to print
[DEBUG] 2017-07-24 17:26:51,752 @ qz.printer.PrintServiceMatcher:?
	Searching for PrintService matching 10.10.0.50
[DEBUG] 2017-07-24 17:26:51,753 @ qz.printer.PrintServiceMatcher:?
	Found 7 printers
[DEBUG] 2017-07-24 17:26:51,761 @ qz.printer.PrintServiceMatcher:?
	Found match: 10.10.0.50
[WARN] 2017-07-24 17:26:51,761 @ qz.printer.PrintOptions:?
	Cannot read  as a double for density, using default
[WARN] 2017-07-24 17:26:51,762 @ qz.printer.PrintOptions:?
	Cannot read  as a valid value for interpolation, using default
[WARN] 2017-07-24 17:26:51,762 @ qz.printer.PrintOptions:?
	Cannot read  as a double for paperThickness, using default
[DEBUG] 2017-07-24 17:26:51,766 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_54, -l]
[DEBUG] 2017-07-24 17:26:51,784 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS _10_10_0_54: 300dpi
[DEBUG] 2017-07-24 17:26:51,785 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Canon_MX700_series, -l]
[DEBUG] 2017-07-24 17:26:51,821 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS Canon_MX700_series: 600dpi
[DEBUG] 2017-07-24 17:26:51,822 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_53, -l]
[DEBUG] 2017-07-24 17:26:51,853 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PRTLASER3, -l]
[DEBUG] 2017-07-24 17:26:51,879 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, _10_10_0_50, -l]
[DEBUG] 2017-07-24 17:26:51,899 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, Brother_PT_2730, -l]
[DEBUG] 2017-07-24 17:26:51,921 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS Brother_PT_2730: 180dpi
[DEBUG] 2017-07-24 17:26:51,922 @ qz.utils.ShellUtilities:?
	Executing: [lpoptions, -p, PRTLASER, -l]
[DEBUG] 2017-07-24 17:26:51,950 @ qz.utils.ShellUtilities:?
	Parsed default density from CUPS PRTLASER: 600dpi
[WARN] 2017-07-24 17:26:51,950 @ qz.printer.PrintOptions:?
	Cannot read null as a double for fallbackDensity, using default
[DEBUG] 2017-07-24 17:26:51,964 @ qz.printer.action.PrintImage:?
	Parsed 1 images for printing
[TRACE] 2017-07-24 17:26:51,976 @ qz.printer.action.PrintPixel:?
	DPI: 600.0	CNV: 72.0
[TRACE] 2017-07-24 17:26:51,977 @ qz.printer.action.PrintPixel:?
	Drawable area: 0.0,0.0:8.5,11.0
[TRACE] 2017-07-24 17:26:51,977 @ qz.printer.action.PrintPixel:?
	[landscape, color, 60000x60000 dphi, (0.0,0.0)->(215.9,279.4)mm]
[INFO] 2017-07-24 17:26:51,989 @ qz.printer.action.PrintPixel:?
	Starting printing (1 copies)
[WARN] 2017-07-24 17:26:51,989 @ qz.printer.action.PrintPixel:?
	Supported printer densities not found
[TRACE] 2017-07-24 17:26:52,035 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[TRACE] 2017-07-24 17:26:52,035 @ qz.printer.action.PrintImage:?
	Requested page 0 for printing
[DEBUG] 2017-07-24 17:26:52,036 @ qz.printer.action.PrintImage:?
	Scaling image up by x8.333333333333334
[DEBUG] 2017-07-24 17:26:52,150 @ qz.printer.action.PrintImage:?
	Paper area: 18,16:754,578
[TRACE] 2017-07-24 17:26:52,150 @ qz.printer.action.PrintImage:?
	Image size: 754.800048828125,301.7387950297307
[TRACE] 2017-07-24 17:26:52,151 @ qz.printer.action.PrintImage:?
	{Fractional metrics enable key=Integer text metrics mode, Image interpolation method key=Bicubic image interpolation mode, Global antialiasing enable key=Antialiased rendering mode, Color rendering quality key=Highest quality color rendering mode, Alpha blending interpolation method key=Highest quality alpha blending methods, Text-specific antialiasing enable key=Default antialiasing text mode, Stroke normalization control key=Pure stroke conversion for accurate paths, Global rendering quality key=Highest quality rendering methods, Text-specific LCD contrast key=140}
[DEBUG] 2017-07-24 17:26:52,153 @ qz.printer.action.PrintImage:?
	Memory: 912m/2731m
[INFO] 2017-07-24 17:26:52,200 @ qz.utils.PrintingUtilities:?
	Printing complete
[TRACE] 2017-07-24 17:26:52,200 @ qz.utils.PrintingUtilities:?
	Returning processor back to pool
```

**Comment by tresf** [comment]
@bberenz is there anything else we can do on Mac to gather more details?  Could this be related to [clearing the printing attributes](https://github.com/qzind/tray/blob/pdffix/src/qz/printer/action/PrintPDF.java#L112) on MacOS only?

**Comment by bberenz** [comment]
>  Could this be related to clearing the printing attributes on MacOS only?

Not exactly, @klabarge was using images in the last test. But this is related to the print attributes, as that is where we [set the page orientation](https://github.com/qzind/tray/blob/pdffix/src/qz/printer/action/PrintPixel.java#L42). 

> is there anything else we can do on Mac to gather more details?

We could try running 
```
System.out.println(Arrays.toString(output.getPrintService().getSupportedAttributeCategories()));
```
in the `PrintPixel.printCopies` method. 
See if it outputs anything different that macOS might be trying to do with printers.

**Comment by bberenz** [comment]
I did some testing on a mac-mini where I forced the print dialog to display before actually sending the document off to the printer. For either landscape orientation the dialog displayed the proper  values: 
![sample](https://user-images.githubusercontent.com/5912715/31252193-6749622a-a9ee-11e7-8b28-cf8496848922.png)
But the page printed exactly the same way for both layouts, so it's likely this is a problem on the Java/OS side of things. Looking around at some of the native print dialogs, there does not seem to be a dedicated "reverse landscape" orientation (only a partially hidden checkbox), possibly meaning this setup option is not supported by the OS itself. Referencing back to issue #7 seems to support this idea as well. 

Getting a list of supported attributes (per my last comment) does not show any that would control the "Reverse page orientation" checkbox in the native dialog. It may be possible for us to perform a flip of the content ourselves in software if we detect both a reverse-landscape setting and the OSX platform. @tresf, thoughts?

**Comment by tresf** [comment]
> flip of the content ourselves in software if we detect both a reverse-landscape setting and the OSX platform. @tresf, thoughts?

I think this is a good option.  I assume it will fail on 2.1 HTML (non-rasterized) though, right?  I slightly remember that being the "black sheep" of printer attribute hacks.

**Comment by bberenz** [comment]
I've fixes for max landscape up on the [reverse-fix](https://github.com/qzind/tray/commits/reverse-fix) branch ready for review.
As for 2.1, we will have to look into what's needed for vector html prints after rebasing.

**Comment by klabarge** [comment]
> I've fixes for macOS landscape up on the reverse-fix branch ready for review.

 @bberenz I tested landscape / reverse landscape from this branch on both macOS and Windows, and the results are identical 👍

**Comment by bberenz** [comment]
The `pdffix` branch actually fixed some of the issues my last commits tried to work around.
45fcd147d6ebf210599f4819e6541951bd89d17b rebases the `reverse-fix` branch off that and moves the work around to what's needed.
@klabarge, can you verify everything still prints correctly on the `reverse-fix` branch please?

**Comment by klabarge** [comment]
Mac and Windows seem to have opposite orientations.

I used PDFWriter for macOS, and Microsoft XPS for Windows.

Mac, reverse-landscape

![image](https://user-images.githubusercontent.com/12505463/32197841-2da6af66-bd9c-11e7-9ec0-3a3a68828ee4.png)

Windows reverse-landscape

![image](https://user-images.githubusercontent.com/12505463/32197947-8facd866-bd9c-11e7-8061-09d786890c0e.png)

**Comment by klabarge** [comment]
I will test to a physical printer to confirm my above results.

**Comment by klabarge** [comment]
I tested this to a physical printer (HP Laserjet M603) and cannot reproduce this issue.

landscape and reverse-landscape printing is identical on Windows, macOS, and Linux.

**Comment by tresf** [comment]
Ok, let's merge it back in, close this issue and get one step closer to 2.0.5. :)

**Comment by klabarge** [comment]
Merged via e602da6
